### PR TITLE
[MLIR] Add SyclRuntimeWrapper 

### PR DESCRIPTION
**Pull request by nbpatel**
This PR is a breakdown of the big PR #65539  which enables intel gpu integration. In this PR we add the code for the sycl runtime wrappers and also the cmake modules to find the dependent components. Integration test will be a follow up PR.

This PR is a joint effort by Nishant Patel & Sang Ik Lee.

### CODE REVIEW

**Comment by nbpatel** [comment]
@grypp @joker-eph

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+#include <cstdio>
+#include <cstdlib>
+#include <stdexcept>
+#include <tuple>
+#include <vector>
+
+#include <CL/sycl.hpp>
+#include <level_zero/ze_api.h>
+#include <map>
+#include <mutex>
```
Comment (Hardcode84) [inline mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp:28]: Most of those includes are not being used and can be removed (map, mutex, vector, atomic)

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+#include <sycl/ext/oneapi/backend/level_zero.hpp>
+
+#ifdef _WIN32
+#define SYCL_RUNTIME_EXPORT __declspec(dllexport)
+#else
+#define SYCL_RUNTIME_EXPORT
+#endif // _WIN32
```
Comment (grypp) [inline mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp:35]: This macro exist in other runtimes too. Maybe we need a common header for all of them. But let's keep this in our mind, we can do it later.

**Review by grypp — COMMENTED** [review]
Looks good overall to me. I left a few comments

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    auto platformName = platform.get_info<sycl::info::platform::name>();
+    bool isLevelZero = platformName.find("Level-Zero") != std::string::npos;
+    if (!isLevelZero)
+      continue;
+
+    syclDevice = platform.get_devices()[0];
+    isDeviceInitialised = true;
+    return syclDevice;
+  }
+    throw std::runtime_error("getDefaultDevice failed");
```
[the remark this answers is not in the document]
Comment (Hardcode84) [reply mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp:68]: Just side comment, SYCL itself uses exceptions to report errors, so it's not possible to disable them completely and that's `catchAll` wrapper is for.

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp (hunk 2 above)
Comment (Hardcode84) [reply mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp:21]: There is a cmake module https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html to generate those macro automatically, what the llvm/mlir position on using it?

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (grypp) [reply mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp:67]: Thanks a lot for the explanation! In this case, using `throw std::runtime ` makes perfect sense to me. It aligns the behavior of this small runtime with SYCL. 

I was wondering if you truly needed exceptions here, and it turns out that yes, you do want them.

**Review by grypp — APPROVED** [review]
This looks good to me! I'm not a SYCL expert, but maybe someone with more expertise could take a quick look here. @Hardcode84 ?

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+extern "C" SYCL_RUNTIME_EXPORT sycl::kernel *
+mgpuModuleGetFunction(ze_module_handle_t module, const char *name) {
+  return catchAll([&]() { return getKernel(module, name); });
+}
+
+extern "C" SYCL_RUNTIME_EXPORT void
+mgpuLaunchKernel(sycl::kernel *kernel, size_t gridX, size_t gridY, size_t gridZ,
+                 size_t blockX, size_t blockY, size_t blockZ,
+                 size_t sharedMemBytes, sycl::queue *queue, void **params,
+                 void **extra, size_t paramsCount) {
```
Comment (Hardcode84) [inline mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp:193]: nit: `extra` is not being used and will result in warning, either do `/*extra*/` or `(void)extra`.

**Review by Hardcode84 — APPROVED** [review]
LGTM

File: mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp (hunk 4 above)
Comment (nbpatel) [reply mlir/lib/ExecutionEngine/SyclRuntimeWrappers.cpp]: done

**Comment by nbpatel** [comment]
Thanks for the feedback and approval.  @grypp can you help me merge this PR ?

**Pull request merged**